src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,665 @@
+            body.encode(buf);
+        }
+        debug!("{:?} Sent data with length: {:?}", self, buf.len());
+        let writer = self.get_stream_mut();
+        match write_all_timeout!(writer, buf, self.timeout_options.write_timeout).await {
+            Ok(()) => Ok(()),
+            Err(err) => {
+                debug!("{:?} Failed to send data: {:?}", self, err);
+                Err(RpcError::InternalError(err.to_string()))
+            }
+        }
+    }
+
+    /// Get the next sequence number.
+    pub fn next_seq(&self) -> u64 {
+        self.seq.fetch_add(1, std::sync::atomic::Ordering::AcqRel)
+    }
+
+    /// Send keep alive message to the server
+    pub async fn ping(&self) -> Result<(), RpcError> {
+        // Send keep alive message
+        let current_seq = self.next_seq();
+        let current_timestamp = self.clock_instant.elapsed().as_secs();
+        // Check keepalive is valid
+        let received_keepalive_timestamp = self
+            .received_keepalive_timestamp
+            .load(std::sync::atomic::Ordering::Acquire);
+        if current_timestamp - received_keepalive_timestamp
+            > self.timeout_options.keep_alive_timeout.as_secs()
+        {
+            debug!("{:?} keep alive timeout, close the connection", self);
+            return Err(RpcError::InternalError("Keep alive timeout".to_owned()));
+        }
+
+        // Set to packet task
+        let keep_alive_msg = ReqHeader {
+            seq: current_seq,
+            op: ReqType::KeepAliveRequest.to_u8(),
+            len: 0,
+        };
+
+        if let Ok(()) = self.send_data(&keep_alive_msg, None).await {
+            debug!("{:?} Success to sent keep alive message", self);
+            Ok(())
+        } else {
+            debug!("{:?} Failed to send keep alive message", self);
+            Err(RpcError::InternalError(
+                "Failed to send keep alive message".to_owned(),
+            ))
+        }
+    }
+
+    /// Send packet by the client.
+    pub async fn send_packet(&self, req_packet: &mut P) -> Result<(), RpcError> {
+        let current_seq = self.next_seq();
+        req_packet.set_seq(current_seq);
+        debug!("{:?} Try to send request: {:?}", self, current_seq);
+
+        // Send may be used in different threads, so we need to create local buffer to store the request data
+        let req_len = req_packet.get_req_len(); // Try to get request buffer
+        let req_header = ReqHeader {
+            seq: req_packet.seq(),
+            op: req_packet.op(),
+            len: req_len,
+        };
+
+        // concate req_header and req_buffer
+        if let Ok(()) = self.send_data(&req_header, Some(req_packet)).await {
+            debug!("{:?} Sent request success: {:?}", self, req_packet.seq());
+            // We have set a copy to keeper and manage the status for the packets keeper
+            // Set to packet task with clone
+            self.packets_keeper.add_task(req_packet)?;
```

Comment:
Move add_task() inside send_data

Reply:
OK, add a `add_task()` function for this client.

Reply:
Maybe we don't want to keep ping message to packets_keeper?

Reply:
In `send_data`, it only have `Encode` trait not the packet trait.

Encode trait is for `header` and `body`, but packets_keeper only keep the body.